metals/src/main/scala/scala/meta/internal/metals/OutlineFilesProvider.scala

```diff
@@ -0,0 +1,191 @@
+package scala.meta.internal.metals
+
+import java.util.Optional
+import java.util.concurrent.atomic.AtomicBoolean
+import java.{util => ju}
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.VirtualFileParams
+import scala.meta.pc.{OutlineFiles => JOutlineFiles}
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+
+class OutlineFilesProvider(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+) {
+  private val outlineFiles =
+    new TrieMap[BuildTargetIdentifier, BuildTargetOutlineFilesProvider]()
+
+  def shouldRestartPc(
+      id: BuildTargetIdentifier,
+      reason: PcRestartReason,
+  ): Boolean = {
+    reason match {
+      case DidCompile(true) => true
+      case _ =>
+        outlineFiles.get(id) match {
+          case Some(provider) =>
+            // if it was never compiled successfully by the build server
+            // we don't restart pc not to lose information from outline compile
+            provider.wasSuccessfullyCompiledByBuildServer
```

Comment:
Is this not a problem if we failed to compile, but it previously did compile? So essentially if we have DidCompile(false) we will restart the compiler and lose the outlined files?

Reply:
No, we will lose it in pc but the changed files are remembered in the provider, so those will be again supplied for re-outline to pc. I just don't want to do it if it was never compiled before, since it will take too long too often.

Reply:
Ok, thanks for explaining, might be worth adding that info too